lib/Feature/FeatureModelWriter.cpp

```diff
@@ -0,0 +1,342 @@
+  if (Writer == nullptr) {
+    return std::nullopt;
+  }
+  RC = writeFeatureModel(Writer.get());
+  if (RC < 0) {
+    xmlFreeDoc(Doc);
+    return std::nullopt;
+  }
+
+  xmlChar *XmlBuff;
+  int Buffersize;
+  xmlDocDumpMemoryEnc(Doc, &XmlBuff, &Buffersize, ENCODING);
+  std::string Str((char *)XmlBuff, Buffersize);
+  xmlFree(XmlBuff);
+
+  // xmlFreeTextWriter called implicitly
+  xmlFreeDoc(Doc);
+  return Str;
+}
+
+int FeatureModelXmlWriter::writeFeatureModel(xmlTextWriterPtr Writer) {
+  int RC;
+
+  RC = xmlTextWriterSetIndent(Writer, 1);
+  CHECK_RC
+  RC = xmlTextWriterSetIndentString(Writer, INDENTATIONSTRING);
+  CHECK_RC
+
+  RC = xmlTextWriterStartDocument(Writer, "1.0", ENCODING, nullptr);
+  CHECK_RC
+  RC = writeVm(Writer);
+
+  return RC;
+}
+
+int FeatureModelXmlWriter::writeVm(xmlTextWriterPtr Writer) {
+  int RC;
+
+  RC = xmlTextWriterStartElement(Writer, XmlConstants::VM);
+  CHECK_RC
+  RC = xmlTextWriterWriteAttribute(Writer, XmlConstants::NAME,
+                                   BAD_CAST Fm.getName().data());
+  CHECK_RC
+  RC = xmlTextWriterWriteAttribute(Writer, XmlConstants::ROOT,
+                                   BAD_CAST Fm.getPath().string().data());
+  CHECK_RC
+
+  RC = writeBinaryFeatures(Writer);
+  CHECK_RC
+
+  RC = writeNumericFeatures(Writer);
+  CHECK_RC
+
+  RC = writeBooleanConstraints(Writer);
+  CHECK_RC
+
+  // TODO mixed and nonNumeric constraints when supported
+
+  RC = xmlTextWriterEndDocument(Writer); // VM
+  return RC;
+}
+
+int FeatureModelXmlWriter::writeBinaryFeatures(xmlTextWriterPtr Writer) {
+  int RC;
+
+  RC = xmlTextWriterStartElement(Writer, XmlConstants::BINARYOPTIONS);
+  CHECK_RC
+
+  for (Feature *F : Fm.features()) {
+    if (F->getKind() == Feature::FeatureKind::FK_BINARY) {
+      RC = writeFeature(Writer, *F);
+      CHECK_RC
+    }
+  }
+
+  RC = xmlTextWriterEndElement(Writer); // BINARYOPTIONS
+  return RC;
+}
+
+int FeatureModelXmlWriter::writeNumericFeatures(xmlTextWriterPtr Writer) {
+  int RC;
+
+  RC = xmlTextWriterStartElement(Writer, XmlConstants::NUMERICOPTIONS);
+  CHECK_RC
+
+  for (auto *F : Fm.features()) {
+    if (F->getKind() == Feature::FeatureKind::FK_NUMERIC) {
+      RC = writeFeature(Writer, *F);
+      CHECK_RC
+    }
+  }
+
+  RC = xmlTextWriterEndElement(Writer); // NUMERICOPTIONS
+  return RC;
+}
+
+int FeatureModelXmlWriter::writeBooleanConstraints(xmlTextWriterPtr Writer) {
+  int RC;
+
+  RC = xmlTextWriterStartElement(Writer, XmlConstants::BOOLEANCONSTRAINTS);
+  CHECK_RC
+
+  // collect alternative groups
+  std::set<std::set<Feature *>> AltGroups;
+  for (auto *F : Fm.features()) {
+    // skip empty groups
+    if (F->alternatives().begin() == F->alternatives().end()) {
+      continue;
+    }
+    std::set<Feature *> AltGroup(F->alternatives().begin(),
+                                 F->alternatives().end());
+    AltGroup.insert(F);
+    AltGroups.insert(AltGroup);
+  };
+
+  for (auto const &Group : AltGroups) {
```

Comment:
```suggestion
  for (const auto &Group : AltGroups) {
```
Similar to llvm we use `const west` style :)